filecoin-proofs/src/bin/phase2.rs

```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
-                    start.elapsed().as_secs()
-                );
-                params_before
-            }
-        };
+    info!(
+        "making streamer from small 'before' params: {}",
+        path_before
+    );
 
-        let params_after = {
-            info!("reading 'after' params from disk: {}", path_after);
-            let file = File::open(path_after).unwrap();
-            let mut reader = BufReader::with_capacity(1024 * 1024, file);
-            let start = Instant::now();
-            let params_after = phase2::MPCParameters::read(&mut reader, true).unwrap();
-            info!(
-                "successfully read 'after' params from disk, dt_read={}s",
-                start.elapsed().as_secs()
+    let mut streamer = if param_size.is_large() {
+        panic!("cannot convert large param format");
+    } else {
+        Streamer::new(path_before, read_raw, write_raw).unwrap_or_else(|e| {
+            panic!(
+                "failed to make streamer from small `{}`: {}",
+                path_before, e
             );
-            params_after
-        };
-
-        info!("verifying param transition");
-        let start_verification = Instant::now();
-
-        let calculated_contribution_hash =
-            phase2::verify_contribution(&params_before, &params_after).expect("invalid transition");
-
-        assert_eq!(
-            &provided_contribution_hash[..],
-            &calculated_contribution_hash[..],
-            "provided contribution hash ({}) does not match calculated contribution hash ({})",
-            hex::encode(&provided_contribution_hash[..]),
-            hex::encode(&calculated_contribution_hash[..]),
-        );
+        })
+    };
 
-        info!(
-            "successfully verified param transition, dt_verify={}s",
-            start_verification.elapsed().as_secs()
+    info!("writing small 'after' params to file: {}", path_after);
+    let file_after = File::create(&path_after).unwrap_or_else(|e| {
+        panic!(
+            "failed to create 'after' params file `{}`: {}",
+            path_after, e
         );
+    });
 
-        next_params_before = Some(params_after);
-    }
+    streamer
+        .process(file_after, CHUNK_SIZE)
+        .expect("failed to convert");
 
     info!(
-        "successfully verified all param transitions, dt_total={}s",
+        "successfully converted, dt_total={}s",
         start_total.elapsed().as_secs()
     );
 }
 
-fn verify_param_transistions_daemon(proof: Proof, hasher: Hasher, sector_size: u64) {
-    const SLEEP_SECS: u64 = 10;
+/// Contributes entropy to the current phase2 parameters for a circuit, then writes the updated
+/// parameters to a new file.
+fn contribute_to_params(path_before: &str) {
+    let (proof, hasher, sector_size, head, prev_param_number, param_size, read_raw) =
+        parse_params_filename(path_before);
+
+    dbg!("in contribute_to_params");
+    assert_eq!(false, read_raw, "Raw format only supported when streaming.");
```

Comment:
This function is the non-streaming contribution, right? `MPCSmall::contribute()` can operate on both raw and non-raw?

Reply:
Yes. This assertion and the FIXME both date to an earlier stage in development when it was not yet possible to create an `MPCSmall` from raw format. I have removed the assertion and addressed the FIXME so this now-possible path is allowed.